### DISCUSSION TITLE
How exactly do the `include` and `exclude` options work?

### DISCUSSION THREAD

**Discussion by anthony-S93** [microsoft/pyright#4829] [Q&A] [answered]
Pyright's [documentation](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options) states that they are used to include/exclude any files and directories in/from the analysis. But what exactly does "analysis" entail?

I tried to understand this via the following setup: 
1) Create a root directory with the following structure
```
root
   \_ main.py
   \_ pyrightconfig.json
   \_ pkg
        \_mod.py
``` 
2) Set up the individual source files as follows
```python
# pkg/mod.py

def foobar(arg: str):
   print(arg)
```
```python
# main.py
from pkg.mod import foobar
foobar(range(0,12)) # try to pass something other than a string into foobar()
```
Without setting the `exclude` option, Pyright reports the mismatch of the argument type as expected. But what surprised me is that the same error is reported even after I added the `exclude` option inside `pyrightconfig.json`

```json
{
    "exclude": [ "./pkg/mod.py", "./main.py" ]
}
```
I am clearly missing something here. Doesn't the `exclude` option exclude files and directories from Pyright's analysis?

Environment
```
pyright version: pyright 1.1.300
editor: nvim v0.8.3
```

**Comment by erictraut** [comment, marked as answer]
The `include`, `exclude` and `ignore` options work the same way in Pyright as they do in the TypeScript compiler.

The `include` and `exclude` options define the "root" files within your workspace that you would like pyright to type check. These files, along with any workspace files that are imported directly or indirectly by these root files comprise your "program". It's not uncommon to want to exclude certain directories from this set of "root" files — for example, a directory of tests or build scripts.

I recreated your example above and ran the pyright CLI on the resulting directory. If you exclude both `main.py` and `pkg/mod.py` and run pyright, I see the following output, as expected:

```
No source files found.
0 errors, 0 warnings, 0 informations 
```

If I exclude only `main.py`, the program now consists of only `pkg/mod.py`, and I receive:
```
Found 1 source file
0 errors, 0 warnings, 0 informations 
```

If I exclude only `pkg/mod.py`, the program now consists of `main.py` (as a root file) and `pkg/mod.py` (as an imported file), and I receive:
```
Found 1 source file
1 error, 0 warnings, 0 informations 
```

If you want to suppress the diagnostic output for a file that would otherwise be included in your "program", you can use the `ignore` option to do so.

**Reply by anthony-S93** [reply]
> If you want to suppress the diagnostic output for a file that would otherwise be included in your "program", you can use the `ignore` option to do so.

I see. I think I got a little confused when I still see diagnostic messages in my LSP after I excluded both files, which is a little counter-intuitive. For example, this is after I used both `exclude` and `ignore` on both files: 
![2023-03-24-110536_1094x129_scrot](https://user-images.githubusercontent.com/69449791/227414022-da862a42-794e-4f61-b2f9-59769fa297fa.png)

On the other hand, using `pyright` on the command line does indeed exclude both files. Is it reasonable in this case to assume that `pyright` wasn't responsible for the diagnostic message in the screenshot but something else (like Neovim's lsp)?

**Reply by erictraut** [reply]
When you open a file in your editor, pyright type checks it even if it's not part of your program.

**Reply by anthony-S93** [reply]
I see. Thanks again for the clarification. 

> When you open a file in your editor, pyright type checks it even if it's not part of your program.

Should this be included in the documentation for clarification?

**Reply by erictraut** [reply]
I don't think there's a logical place to put it in the documentation. It's pretty standard behavior for a language server to act upon any open file that corresponds to its supported language, so I think it would be surprising to most users if pyright didn't do this.
